File: examples/java/ExampleDotProduct.java
import yacx.Executor;
import ya


public class ExampleDotProduct {
    public static void main(String[] args ){
        //load library
        Executor.load();

        //testdata
        final int numThreads = 4;
        final int numBlocks = 1;

        int n = numThreads * numBlocks;
        float[] x = new float[n];
        float[] y = new float[n];

        for(int i = 0; i < n; i++){
            x[i] = i;
            y[i] = 2 * i;
        }

        //initalize arguments
        KernelArg nArg, xArg, yArg;
        FloatArg outArg;
        xArg = FloatArg.create(x);
        yArg = FloatArg.create(y);
        outArg = FloatArg.createOutput(n);
        nArg = IntArg.createValue(n);

        // Create Program
        String kernelString = Utils.loadFile("kernels/dotProduct.cu");
        Program dotProduct = Program.create(kernelString, "dotProduct");

        // Create compiled Kernel
        Kernel dotProductKernel = dotProduct.compile();

        // Launch Kernel
        KernelTime executionTime = dotProductKernel.launch(numThreads, numBlocks, xArg, yArg, outArg, nArg);

        // Get Result
        float[] out = outArg.asFloatArray();

        // Print Result
        System.out.println("\ndotProduct-Kernel sucessfully launched:");
        System.out.println(executionTime);
        System.out.println("\nInput x: " + Arrays.toString(x));
        System.out.println("Input y: " + Arrays.toString(y));
        System.out.println("Result:  " + Arrays.toString(out));


    }



}

File: examples/scala/ExampleDotProduct.scala
import java.io.IOException
import java.util.Arrays
import yacx.Executor
import yacx.FloatArg
import yacx.KernelTime

object ExampleDotProduct {
  def main(args: Array[String]) : Unit = {
    // Load library
    Executor.loadLibrary()

    // Testdata
    val numberElements = 9
    val x = new Array[Float](numberElements)
    val y = new Array[Float](numberElements)
    for (i <- 0 until numberElements) {
      x(i) = i
      y(i) = 2 * i
    }

    // Initalize arguments
    val xArg = FloatArg.create(x: _*)
    val yArg = FloatArg.create(y: _*)
    val outArg = FloatArg.createOutput(1)

    // Compile and Launch
    val executionTime = Executor.launch("dotProduct", 1, numberElements, xArg, yArg, outArg)

    // Get Result
    val result = outArg.asFloatArray()(0)

    // Print Result
    println("\ndotProduct-Kernel sucessfully launched:")
    println(executionTime)
    println("\nInput x: " + Arrays.toString(x))
    println("Input y: " + Arrays.toString(y))
    println("Result:  " + result)
  }
}
